add `CStr` method that accepts any slice containing a nul-terminated string

ericseppanen opened the pull request:
I haven't created an issue (tracking or otherwise) for this yet; apologies if my approach isn't correct. This is my first code contribution.

This change adds a member fn that converts a slice into a `CStr`; it is intended to be safer than `from_ptr` (which is unsafe and may read out of bounds), and more useful than `from_bytes_with_nul` (which requires that the caller already know where the nul byte is).

The reason I find this useful is for situations like this:
```rust
let mut buffer = [0u8; 32];
unsafe {
    some_c_function(buffer.as_mut_ptr(), buffer.len());
}
let result = CStr::from_bytes_with_nul(&buffer).unwrap();
```

This code above returns an error with `kind = InteriorNul`, because `from_bytes_with_nul` expects that the caller has passed in a slice with the NUL byte at the end of the slice. But if I just got back a nul-terminated string from some FFI function, I probably don't know where the NUL byte is.

I would wish for a `CStr` constructor with the following properties:
- Accept `&[u8]` as input
- Scan for the first NUL byte and return the `CStr` that spans the correct sub-slice (see [future note below](https://github.com/rust-lang/rust/pull/94984#issuecomment-1070754281)).
- Return an error if no NUL byte is found within the input slice

I asked on [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/122651-general/topic/CStr.20from.20.26.5Bu8.5D.20without.20knowing.20the.20NUL.20location.3F) whether this sounded like a good idea, and got a couple of positive-sounding responses from @joshtriplett and @AzureMarker.

This is my first draft, so feedback is welcome.

A few issues that definitely need feedback:

1. Naming. @joshtriplett called this `from_bytes_with_internal_nul` on Zulip, but after staring at all of the available methods, I believe that this function is probably what end users want (rather than the existing fn `from_bytes_with_nul`). Giving it a simpler name (**`from_bytes`**) implies that this should be their first choice.
2. Should I add a similar method on `CString` that accepts `Vec<u8>`? I'd assume the answer is probably yes, but I figured I'd try to get early feedback before making this change bigger.
3. What should the error type look like? I made a unit struct since `CStr::from_bytes` can only fail in one obvious way, but if I need to do this for `CString` as well then that one may want to return `FromVecWithNulError`. And maybe that should dictate the shape of the `CStr` error type also?

Also, cc @poliorcetics who wrote #73139 containing similar fns.

rust-highfive commented:
r? @Mark-Simulacrum

(rust-highfive has picked a reviewer for you, use r? to override)

AzureMarker commented on library/std/src/ffi/c_str/tests.rs:
```diff
@@ -117,6 +117,43 @@ fn from_bytes_with_nul_interior() {
+    // Test an empty slice. This should fail because it
+    // does not contain a nul byte.
+    let b = b"";
+    CStr::from_bytes(&b[..]).unwrap_err();
```
Instead of unwraps I would expect these to be `assert_eq` macro calls.

ericseppanen replied:
Thanks @AzureMarker for the feedback.

For the error cases, sure, I can replace `unwrap_err` with `assert_eq!(r, Err(FromBytesError))`. It's a unit type right now, but I can see how it sets a good example (also the error type may change).

If your concern is also about the success cases, I don't see any way to do equality comparisons on a `CStr`. Each of these are equivalent for my purposes:
```
assert_eq!(r.to_str().unwrap(), "hello");   // What I'm currently doing
assert_eq!(r.to_string_lossy(), "hello");   // Creates a Cow, which seems a little awkward
assert_eq!(r.to_bytes(), b"hello");         // Maybe this would be best?
```

AzureMarker replied:
Yeah, the success cases would have to be a little different. I think your byte string example would work best.

ericseppanen replied:
`to_bytes` omits the `\0`; there's also `to_bytes_with_nul` :)

ericseppanen replied:
There's still another `unwrap()` in the success cases, e.g.
```
let b = b"hello\0world!";
let r = CStr::from_bytes(&b[..]).unwrap();
assert_eq!(r.to_bytes(), b"hello");
```

I'm not sure there's a way to remove the unwrap without making the test code a lot less readable. The other tests in this file all do similar things, FWIW.

AzureMarker replied:
That should be fine.

m-ou-se commented on library/std/src/ffi/c_str.rs:
```diff
@@ -328,6 +328,17 @@ impl FromVecWithNulError {
+///
+#[derive(Clone, PartialEq, Eq, Debug)]
+#[unstable(feature = "tolerant_cstr_from_bytes", issue = "none")]
+pub struct FromBytesError;
```
If you define this without any private members, anyone can create this struct, which we usually prevent to make sure we can change things in the future. Check some of the other error types in std, and you'll see some have a single private `()` member because of this.

And unrelated to that: error types should implement `Error` (and therefore also `Display`).

AzureMarker replied:
Another way of doing this is to use `#[non_exhaustive]`

m-ou-se commented:
I think this should have a more detailed name than `from_bytes`. It's not clear from the name that the result will not include all the bytes, but only the part until the first nul byte. `from_bytes_until_nul` maybe?

poliorcetics commented:
I agree with @m-ou-se, especially since this is mostly intended for FFI, a very clear name is better.

The alternative for CString would be nice too, but that can be done in a separate PR (and marked as TODO in the tracking issue)

You said you didn't know how to create a tracking issue. For this you can simply copy an existing one and rework its informations to be about this PR :)

m-ou-se commented:
> You said you didn't know how to create a tracking issue. For this you can simply copy an existing one and rework its informations to be about this PR :)

We have [a template](https://github.com/rust-lang/rust/issues/new?assignees=&labels=C-tracking-issue%2C+T-libs&template=library_tracking_issue.md&title=Tracking+Issue+for+XXX) for this. :)

ericseppanen commented:
The tracking issue template says
> If the new feature is small, it may be fine to skip the RFC process. In that
> case, you can use use `issue = "none"` in your initial implementation PR. The
> reviewer will ask you to open a tracking issue if they agree your feature can be
> added without an RFC.

Can I take the above comments as approval to go ahead and create one now?

ericseppanen commented:
Updated since the first draft:
- renamed to `from_bytes_until_nul` (and `FromBytesUntilNulError`)
- fixed Error type (impl `Error` + `Display`, add private `()` member)
- improved docs

I think this is ready for review.

mbartlett21 commented on library/std/src/ffi/c_str.rs:
```diff
@@ -328,6 +328,27 @@ impl FromVecWithNulError {
+#[unstable(feature = "cstr_from_bytes_until_nul", issue = "95027")]
+impl fmt::Display for FromBytesUntilNulError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "data provided is not nul terminated")
```
This seems to imply that it should end with a nul.

Perhaps `data provided does not contain a nul`?

ericseppanen replied:
I agree that sounds better.

mbartlett21 commented on library/std/src/ffi/c_str.rs:
```diff
@@ -1239,12 +1260,60 @@ impl CStr {
     /// wrapper after ensuring that the byte slice is nul-terminated
     /// and does not contain any interior nul bytes.
     ///
+    /// If the caller does not know where the nul byte is located,
```
```suggestion
    /// If the nul byte may not be at the end,
```

m-ou-se commented:
> the `CStr` that **spans** the correct sub-slice

It's important to note that while `&CStr` is currently a wide pointer (pointer+length), the plan is to change that to be just a single pointer, like a `const char *` in C, as soon as we have the language features to do so. When that happens, `from_ptr` will become O(1), and `len` will become O(N). Then, this `from_bytes_until_nul` will basically become a version of the trivial `from_ptr` that in addition scans if there's really a terminating nul byte, but does *not* return that length. Calling `len` after `from_bytes_until_nul` will have to scan the bytes again.

Mark-Simulacrum commented:
r=me with commits squashed into one. I also updated the tracking issue with @m-ou-se's concern.

ericseppanen commented:
> I also updated the tracking issue with @m-ou-se's concern.

I had read that message as informative, just warning that the word "span" may no longer apply in the future. Can you clarify what is the remaining concern?

Even in a future where `CStr` is just a pointer, I think the need for this function still exists (because `from_bytes_with_nul` would still return an error if the nul isn't at the end of the slice).

m-ou-se commented:
Once we 'fix' &CStr, this function basically just does a strlen() and then throws away the result. I'm not saying it'd be useless, but it doesn't fit as nicely into the API as it does today. So we should at least think about that a bit more before stabilizing.

ericseppanen commented:
> Once we 'fix' &CStr, this function basically just does a strlen() and then throws away the result. I'm not saying it'd be useless, but it doesn't fit as nicely into the API as it does today. So we should at least think about that a bit more before stabilizing.

We don't need the length, we need the safety (a guarantee that there must be a null byte in that slice somewhere). I don't see any alternative.

ericseppanen commented:
@Mark-Simulacrum I pushed a rebased + squashed version.

(Apologies, the rebase makes github's "compare" broken-- I didn't realize that would happen.)

Mark-Simulacrum commented:
@bors r+ rollup

Yes, rebase fixing compare is fine and expected.

bors commented:
:pushpin: Commit d5fe4cad5a68a305b9fa7d421b0ffa358b14b0a9 has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum d5fe4cad5a68a305b9fa7d421b0ffa358b14b0a9 -->
<!-- homu: {"type":"Approved","sha":"d5fe4cad5a68a305b9fa7d421b0ffa358b14b0a9","approver":"Mark-Simulacrum"} -->
